internal/tools/kops/instance_groups.go

```diff
@@ -20,7 +22,80 @@ type InstanceGroupMetadata struct {
 
 // InstanceGroupSpec is the spec of a kops instance group.
 type InstanceGroupSpec struct {
-	Image string `json:"image"`
+	Role        string `json:"role"`
+	Image       string `json:"image"`
+	MachineType string `json:"machineType"`
+	MinSize     int64  `json:"minSize"`
+	MaxSize     int64  `json:"maxSize"`
+}
+
+// UpdateMetadata updates KopsMetadata with the current values from kops state
+// store. This can be a bit tricky. We are attempting to correlate multiple kops
+// instance groups into a simplified set of metadata information. To do so, we
+// assume and check the following:
+// - There is one worker node instance group.
+// - There is one or more master instance groups.
+// - All of the cluster hosts are running the same AMI.
+// - All of the master nodes are running the same instance type.
+// Note:
+// If any violations are found, we don't return an error as that is beyond the
+// scope of updating the metadata. Instead, warnings for each violation are
+// returned and stored.
+func (c *Cmd) UpdateMetadata(metadata *model.KopsMetadata) error {
+	instanceGroups, err := c.GetInstanceGroupsJSON(metadata.Name)
+	if err != nil {
+		return err
+	}
+
+	var masterIGCount, NodeIGCount, nodeMinCount, nodeMaxCount int64
+	var masterMachineType, nodeInstanceType, AMI string
+	for _, ig := range instanceGroups {
+		switch ig.Spec.Role {
+		case "Master":
+			if AMI == "" {
+				AMI = ig.Spec.Image
+			} else if AMI != ig.Spec.Image {
+				metadata.AddWarning(fmt.Sprintf("Expected all hosts to be running same AMI, but instance group %s is has AMI %s", ig.Metadata.Name, ig.Spec.Image))
+			}
+
+			if masterMachineType == "" {
+				masterMachineType = ig.Spec.MachineType
+			} else if masterMachineType != ig.Spec.MachineType {
+				metadata.AddWarning(fmt.Sprintf("Expected all master hosts to be running same machine type, but instance group %s is has type %s", ig.Metadata.Name, ig.Spec.MachineType))
+			}
+
+			masterIGCount++
+		case "Node":
+			if AMI == "" {
+				AMI = ig.Spec.Image
+			} else if AMI != ig.Spec.Image {
+				metadata.AddWarning(fmt.Sprintf("Expected all hosts to be running same AMI, but instance group %s is has AMI %s", ig.Metadata.Name, ig.Spec.Image))
+			}
+
+			NodeIGCount++
+			nodeInstanceType = ig.Spec.MachineType
+			nodeMinCount = ig.Spec.MinSize
+			nodeMaxCount = ig.Spec.MaxSize
+		default:
+			metadata.AddWarning(fmt.Sprintf("Instance group %s has unknown role %s", ig.Metadata.Name, ig.Spec.Role))
+		}
+	}
+
+	if masterIGCount == 0 {
+		metadata.AddWarning("Failed to find any master instance groups")
+	}
+	if NodeIGCount != 1 {
+		metadata.AddWarning(fmt.Sprintf("expected exactly 1 node instance group, but found %d", NodeIGCount))
```

Comment:
Interesting approach with storing the warnings with the metadata. If I'm reading this right we don't log the warnings any where which would mean the only way we'd notice them is if looked at the cluster itself, right? I imagine you want to store them as they're technically part of the kops state? Should we log them or set up some grafana alerts on the cluster table so we are aware of them?

Reply:
Yeah, that's exactly right: the warnings are stored with the metadata if there are any. I initially started with logging the warnings, but found that having them presented with the cluster metadata worked really well. Basically, you can list your clusters and know that any kops validation errors will be right there if there are any. On the flip side, you know that no errors means validation was clean.

All that said, there is absolutely no reason we can't have our cake and eat it too so I just pushed up a change to log the warnings as well. Great idea on automated alerts!